### PR TITLE
Add p2tr keyspend finalization support for PSBT

### PR DESCRIPTION
**Pull request by instagibbs**
Chunking off another small piece of keyspend support.

### CODE REVIEW

File: src/psbt.c [hunk 1]
```diff
@@ -4405,6 +4405,24 @@ static bool finalize_multisig(struct wally_psbt_input *input,
...
+
+    /* TODO support tapleaf spends input->taproot_leaf_signatures */
+    if (!sig ||
+        wally_witness_keyspend_p2tr(sig->value, sig->value_len,
+                                    &final_witness) != WALLY_OK)
+        return false;
+
+    if (wally_psbt_input_set_final_witness(input, final_witness) != WALLY_OK)
+        return false;
+    return true;
```
Comment (jgriffiths) [inline src/psbt.c:4423]: as above, if you generate directly you don't need to set this here.

File: src/wasm_package/src/const.js [hunk 2]
```diff
@@ -121,10 +121,12 @@ export const WALLY_MINISCRIPT_ONLY = 0x02; /** Only allow miniscript (not descri
 export const WALLY_MINISCRIPT_REQUIRE_CHECKSUM = 0x04; /** Require a checksum to be present */
 export const WALLY_MINISCRIPT_TAPSCRIPT = 0x01; /** Tapscript, use x-only pubkeys */
 export const WALLY_MS_CANONICAL_NO_CHECKSUM = 0x01; /** Do not include a checksum */
-export const WALLY_MS_IS_DESCRIPTOR = 0x08; /** Contains only descriptor expressions (no miniscript) */
+export const WALLY_MS_IS_DESCRIPTOR = 0x20; /** Contains only descriptor expressions (no miniscript) */
```
[the remark this answers is not in the document]
Comment (jgriffiths) [reply src/wasm_package/src/const.js:124]: edit: master is now updated, please rebase to it ta

[the remark this answers is not in the document]
Comment (instagibbs) [reply src/wasm_package/src/const.js:124]: wondered about that, thanks

File: src/script.c [hunk 3]
```diff
@@ -1235,6 +1235,28 @@ int wally_witness_p2wpkh_from_sig(
...
+    size_t sig_len,
+    struct wally_tx_witness_stack **witness)
+{
+    int ret;
+
+    /* Required to be a valid BIP340 length of 64 + possible sighash flag */
+    if (sig_len != 64 && sig_len != 65)
+        return WALLY_EINVAL;
+
+    if (witness)
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/script.c:1245]: done

File: src/script.c [hunk 4]
```diff
@@ -1235,6 +1235,28 @@ int wally_witness_p2wpkh_from_sig(
...
 
+int wally_witness_keyspend_p2tr(
+    const unsigned char *sig,
+    size_t sig_len,
+    struct wally_tx_witness_stack **witness)
+{
+    int ret;
+
+    /* Required to be a valid BIP340 length of 64 + possible sighash flag */
+    if (sig_len != 64 && sig_len != 65)
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/script.c]: done

File: src/script.c [hunk 5]
```diff
@@ -1235,6 +1235,28 @@ int wally_witness_p2wpkh_from_sig(
...
+        return WALLY_EINVAL;
+
+    if (witness)
+        *witness = NULL;
+
+    ret = wally_tx_witness_stack_init_alloc(1, witness);
+
+    if (ret == WALLY_OK)
+        return wally_tx_witness_stack_add(*witness, sig, sig_len);
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/script.c:1255]: done

File: include/wally_script.h [hunk 6]
```diff
@@ -293,6 +293,19 @@ WALLY_CORE_API int wally_witness_p2wpkh_from_der(
...
 
+/**
+ * Create a P2TR keyspend witness from a xonly pubkey and BIP340 signature plus sighash.
+ *
+ * :param sig: The BIP340-encoded signature to create a witness,
+ *|    with the sighash byte appended to it if non-SIGHASH_DEFAULT.
+ * :param sig_len: The length of ``sig`` in bytes. Must be 64 or 65.
+ * :param witness: Destination for the newly created witness.
+ */
+WALLY_CORE_API int wally_witness_keyspend_p2tr(
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply include/wally_script.h]: took wally_witness_p2tr_from_sig

File: include/wally_script.h [hunk 7]
```diff
@@ -293,6 +293,19 @@ WALLY_CORE_API int wally_witness_p2wpkh_from_der(
     size_t sig_len,
     struct wally_tx_witness_stack **witness);
 
+/**
+ * Create a P2TR keyspend witness from a xonly pubkey and BIP340 signature plus sighash.
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply include/wally_script.h]: heh, yes. cleaned up description

File: src/psbt.c [hunk 8]
```diff
@@ -4405,6 +4405,24 @@ static bool finalize_multisig(struct wally_psbt_input *input,
...
+{
+    struct wally_tx_witness_stack *final_witness;
+    const struct wally_map_item *sig;
+
+    sig = wally_map_get_integer(&input->psbt_fields, PSBT_IN_TAP_KEY_SIG);
+
+    /* TODO support tapleaf spends input->taproot_leaf_signatures */
+    if (!sig ||
+        wally_witness_keyspend_p2tr(sig->value, sig->value_len,
+                                    &final_witness) != WALLY_OK)
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/psbt.c]: done

File: src/psbt.c (hunk 1 above)
Comment (instagibbs) [reply src/psbt.c:4420]: done

File: include/wally_script.h [hunk 9]
```diff
@@ -293,6 +293,20 @@ WALLY_CORE_API int wally_witness_p2wpkh_from_der(
     size_t sig_len,
     struct wally_tx_witness_stack **witness);
 
+/**
+ * Create a P2TR keyspend witness from a BIP340 signature plus
+ * optional sighash.
+ *
+ * :param sig: The BIP340-encoded signature to create a witness,
+ *|    with the sighash byte appended to it if non-SIGHASH_DEFAULT.
```
Comment (jgriffiths) [inline include/wally_script.h:301]: suggest:
```
 * :param sig: The BIP340-encoded keyspend signature, including a sighash byte
 *|    for non `WALLY_SIGHASH_DEFAULT` sighashes.
```

File: src/script.c [hunk 10]
```diff
@@ -1235,6 +1235,32 @@ int wally_witness_p2wpkh_from_sig(
     return ret;
 }
 
+int wally_witness_p2tr_from_sig(
+    const unsigned char *sig,
+    size_t sig_len,
+    struct wally_tx_witness_stack **witness)
+{
```
Comment (jgriffiths) [inline src/script.c:1242]: nit: prefer:
```
int wally_witness_p2tr_from_sig(const unsigned char *sig, size_t sig_len,
                                struct wally_tx_witness_stack **witness)
```
The one-line per parameter format is for the headers, in the impls I like to preseve vertical space for easier reviewing if possible.

File: src/script.c [hunk 11]
```diff
@@ -1235,6 +1235,32 @@ int wally_witness_p2wpkh_from_sig(
...
+        return WALLY_EINVAL;
+
+    ret = wally_tx_witness_stack_init_alloc(1, witness);
+
+    if (ret == WALLY_OK) {
+        return wally_tx_witness_stack_add(*witness, sig, sig_len);
+    } else {
+        wally_tx_witness_stack_free(*witness);
+        *witness = NULL;
+    }
```
Comment (jgriffiths) [inline src/script.c:1259]: Almost! :)
If alloc fails there's nothing to free, all wally calls clean up after themselves (although cleaning up again is harmless). You need to catch the case where alloc is ok, but add fails, i.e. something like:

```
    if ((ret = wally_tx_witness_stack_init_alloc(1, witness)) == WALLY_OK) {
        ret = wally_tx_witness_stack_add(*witness, sig, sig_len);
        if (ret != WALLY_OK) {
            wally_tx_witness_stack_free(*witness);
            *witness = NULL;
        }
    }
```

File: src/test/util.py [hunk 12]
```diff
@@ -690,6 +690,7 @@ class wally_psbt(Structure):
     ('wally_wif_to_bytes', c_int, [c_char_p, c_uint32, c_uint32, c_void_p, c_size_t]),
     ('wally_wif_to_public_key', c_int, [c_char_p, c_uint32, c_void_p, c_size_t, c_size_t_p]),
     ('wally_witness_multisig_from_bytes', c_int, [c_void_p, c_size_t, c_void_p, c_size_t, POINTER(c_uint32), c_size_t, c_uint32, POINTER(POINTER(wally_tx_witness_stack))]),
+    ('wally_witness_p2tr_from_sig', c_int, [c_void_p, c_size_t, POINTER(POINTER(wally_tx_witness_stack))]),
```
Comment (jgriffiths) [inline src/test/util.py:693]: Java isn't fully automated yet, for new funcs you need to add an entry to `src/swig_java/swig.i`. For this PR that would look like:
```
diff --git a/src/swig_java/swig.i b/src/swig_java/swig.i
index 580dee20..32c45dc7 100644
--- a/src/swig_java/swig.i
+++ b/src/swig_java/swig.i
@@ -918,6 +918,7 @@ static jobjectArray create_jstringArray(JNIEnv *jenv, char **p, size_t len) {
 %returns_size_t(wally_scriptsig_p2pkh_from_sig);
 %returns_size_t(wally_scriptsig_p2pkh_from_der);
 %returns_size_t(wally_scriptsig_multisig_from_bytes);
+%returns_struct(wally_witness_p2tr_from_sig, wally_tx_witness_stack);
 %returns_struct(wally_witness_p2wpkh_from_sig, wally_tx_witness_stack);
 %returns_struct(wally_witness_p2wpkh_from_der, wally_tx_witness_stack);
 %returns_struct(wally_witness_multisig_from_bytes, wally_tx_witness_stack);
```

**Review by jgriffiths — COMMENTED** [review]
Looks good, I can merge once the above are addressed thanks.

**Comment by instagibbs** [comment]
took all suggestions, thanks

File: src/secp256k1
Comment (jgriffiths) [inline src/secp256k1:1]: You've reverted the secp update here, this needs removing from the PR

Comment (jgriffiths) [reply src/secp256k1:1]: Removed the secp downgrade, added `psbt:` to the commit description and merged, thanks @instagibbs

**Comment by jgriffiths** [comment]
Merged with minor update, closing

**Pull request closed without merging**